### PR TITLE
PSD2 and basic photo tan support

### PR DESCRIPTION
**Pull request by mathiasarens**
Hi Frederick, could you please publish this new version to npmjs?

### CODE REVIEW

**Comment by Prior99** [comment]
I will see if I can perform a review tomorrow and then publish it.

File: packages/fints/README.md [hunk 1]
```diff
@@ -1,13 +1,16 @@
 # fints
 
-[![npm](https://img.shields.io/npm/v/fints.svg)](https://www.npmjs.com/package/fints)
-[![pipeline status](https://gitlab.com/prior99/fints/badges/master/pipeline.svg)](https://github.com/Prior99/fints)
-[![coverage report](https://gitlab.com/prior99/fints/badges/master/coverage.svg)](https://github.com/Prior99/fints)
+[![npm](https://img.shields.io/npm/v/node-fints.svg)](https://www.npmjs.com/package/node-fints)
 
 A client library for communicating with [FinTS servers](https://www.hbci-zka.de/).
 
+Fork from [fints](https://github.com/Prior99/fints)
```
Comment (Prior99) [inline packages/fints/README.md]: Remove these please.

File: packages/fints/README.md [hunk 2]
```diff
@@ -28,6 +31,72 @@ const statements = await client.statements(accounts[0], startDate, endDate);
 console.info(statements); // List of all statements with transactions in specified date range.
 ```
 
+### JavaScript Example
+```javascript
+function getAccountStatements( client, accounts, statements, callback ) {
+  for ( let i in accounts ) {
+    let account = accounts[i]
+
+    if ( !account.hasOwnProperty( "statementsIsSet" ) || (account.hasOwnProperty( "statementsIsSet" ) && !account.statementsIsSet) ) {
```
Comment (Prior99) [inline packages/fints/README.md]: I would like to keep the README small. There is an example CLI in the packages.

File: packages/fints/README.md [hunk 3]
```diff
@@ -28,6 +31,72 @@ const statements = await client.statements(accounts[0], startDate, endDate);
...
+## nodjs Service Example
+Example Application with nodejs middleware and web application for view banking transactions.
+
+###Nodejs application for the fints data processing 
+- [cf-banking-fints-example](https://github.com/guidoMueller/cf-banking-fints-example)
+
+- [DEMO](https://cf-banking-fints-example.cfapps.eu10.hana.ondemand.com/)
+
+###IOS Application work with the nodejs application
+- [IOS Application](https://apps.apple.com/us/app/unibanking/id1469203913)
```
Comment (Prior99) [inline packages/fints/README.md]: Is this an example or an app that uses this library?

File: packages/fints/package-lock.json [hunk 4]
```diff
@@ -0,0 +1,6065 @@
+{
```
Comment (Prior99) [inline packages/fints/package-lock.json]: Yarn is used. Please remove this file.

File: packages/fints/src/__tests__/test-mt940-86-structured.ts [hunk 5]
```diff
@@ -47,7 +47,7 @@ const testData = [
             iban: "DE32250400660167600600",
             reference: {
                 customerRef: "1053050069-20180829182432",
-                date: new Date("2018-08-29T18:25:00.000Z"),
+                date: new Date("2018-08-29T18:25:00.000"),
```
Comment (Prior99) [inline packages/fints/src/__tests__/test-mt940-86-structured.ts]: Please add the `Z`. Without it, this time is local and the tests will behave depending on the timezone of the machine.

File: packages/fints/src/__tests__/test-pin-tan-client-e2e.ts [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+import { TanRequiredError } from "../errors/tan-required-error";
+import { PinTanClient } from "../pin-tan-client";
+import { SEPAAccount } from "../types";
+
+const url = process.env.FINTS_URL;
+const name = process.env.FINTS_USER;
+const pin = process.env.FINTS_PASSWORD;
+const blz = process.env.FINTS_BLZ;
+
+// const url = process.env.COMDIRECT_FINTS_URL;
```
Comment (Prior99) [inline packages/fints/src/__tests__/test-pin-tan-client-e2e.ts]: Please remove the commented-out lines.

File: packages/fints/src/__tests__/test-pin-tan-client-e2e.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+// const blz = process.env.COMDIRECT_FINTS_BLZ;
+
+// const url = process.env.ING_DIBA_FINTS_URL;
+// const name = process.env.ING_DIBA_FINTS_USER;
+// const pin = process.env.ING_DIBA_FINTS_PASSWORD;
+// const blz = process.env.ING_DIBA_FINTS_BLZ;
+
+const productId = "9FA6681DEC0CF3046BFC2F8A6";
+
+test.skip("get accounts", async () => {
```
Comment (Prior99) [inline packages/fints/src/__tests__/test-pin-tan-client-e2e.ts]: Either delete these tests or the `.skip`.

File: packages/fints/src/__tests__/test-pin-tan-client-e2e.ts [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    const client = new PinTanClient({ blz, name, pin, url, productId, debug: true });
+    try {
+        const accounts = await client.accounts();
+        console.info(accounts); // List of all accounts
```
Comment (Prior99) [inline packages/fints/src/__tests__/test-pin-tan-client-e2e.ts]: Please remove this debug print.

File: packages/fints/src/__tests__/test-pin-tan-client-e2e.ts [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        await fs.writeFileSync('/tmp/account.json', JSON.stringify(accounts[0]))
```
Comment (Prior99) [inline packages/fints/src/__tests__/test-pin-tan-client-e2e.ts]: Instead of writing this to a file without an assertion, how about a snapshot test?

File: packages/fints/src/__tests__/test-pin-tan-client-e2e.ts [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+test.skip("get statements", async () => {
+    const client = new PinTanClient({ blz, name, pin, url, productId, debug: true });
+    const account: SEPAAccount = JSON.parse((await fs.readFileSync('/tmp/account.json') as Buffer).toString());
+    const startDate = new Date("2019-09-27T12:00:00Z");
+    const endDate = new Date("2019-12-27T12:00:00Z");
+
+    try {
+        const statements = await client.statements(account, startDate, endDate);
+        console.info(statements);
```
Comment (Prior99) [inline packages/fints/src/__tests__/test-pin-tan-client-e2e.ts]: Please remove this line.

File: packages/fints/src/__tests__/test-pin-tan-client-e2e.ts [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+            fs.writeFileSync('/tmp/challenge.png', error.challengeMedia);
+        } else {
+            console.error(error);
+        }
+    }
+}, 600000);
+
+test.skip("get statements", async () => {
+    const client = new PinTanClient({ blz, name, pin, url, productId, debug: true });
+    const account: SEPAAccount = JSON.parse((await fs.readFileSync('/tmp/account.json') as Buffer).toString());
```
Comment (Prior99) [inline packages/fints/src/__tests__/test-pin-tan-client-e2e.ts]: If this test dependens on the above test, perhaps move the setup into a `beforeEach` block.

File: packages/fints/src/__tests__/test-pin-tan-client-e2e.ts [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+    const startDate = new Date("2019-09-27T12:00:00Z");
+    const endDate = new Date("2019-12-27T12:00:00Z");
+
+    try {
+        const statements = await client.statements(account, startDate, endDate);
+        console.info(statements);
+        await fs.unlinkSync('/tmp/statements-status.txt');
```
Comment (Prior99) [inline packages/fints/src/__tests__/test-pin-tan-client-e2e.ts]: I don't understand the purpose of this test.

File: packages/fints/src/__tests__/test-pin-tan-client-e2e.ts [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+test.skip("complete statements", async () => {
+    const client = new PinTanClient({ blz, name, pin, url, productId, debug: true });
+    const tan: string = '492857';
+
+    try {
+        const tanRequiredError = JSON.parse((await fs.readFileSync('/tmp/statements-status.txt') as Buffer).toString()) as TanRequiredError;
+
+        const statements = await client.completeStatements(tanRequiredError.dialog, tanRequiredError.transactionReference, tan);
+        console.info(statements);
+        await fs.unlinkSync('/tmp/statements-status.txt');
```
Comment (Prior99) [inline packages/fints/src/__tests__/test-pin-tan-client-e2e.ts]: I don't understand the purpose of this test.

File: packages/fints/src/__tests__/test-pin-tan-client-e2e.ts [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+
+test.skip("complete statements", async () => {
+    const client = new PinTanClient({ blz, name, pin, url, productId, debug: true });
+    const tan: string = '492857';
+
+    try {
+        const tanRequiredError = JSON.parse((await fs.readFileSync('/tmp/statements-status.txt') as Buffer).toString()) as TanRequiredError;
+
+        const statements = await client.completeStatements(tanRequiredError.dialog, tanRequiredError.transactionReference, tan);
+        console.info(statements);
```
Comment (Prior99) [inline packages/fints/src/__tests__/test-pin-tan-client-e2e.ts]: Delete this.

File: packages/fints/src/__tests__/test-pin-tan-client.ts [hunk 15]
```diff
@@ -35,7 +36,7 @@ test("accounts", async () => {
     mock.restore();
 });
 
-test("statements", async () => {
+test.skip("statements", async () => {
```
Comment (Prior99) [inline packages/fints/src/__tests__/test-pin-tan-client.ts]: Please remove this `skip`.

File: packages/fints/src/__tests__/test-pin-tan-client.ts [hunk 16]
```diff
@@ -57,7 +58,7 @@ test("statements", async () => {
     mock.restore();
 });
 
-test("balance", async () => {
+test.skip("balance", async () => {
```
Comment (Prior99) [inline packages/fints/src/__tests__/test-pin-tan-client.ts]: Please remove this `skip`.

File: packages/fints/src/__tests__/test-pin-tan-client.ts [hunk 17]
```diff
@@ -79,7 +80,7 @@ test("balance", async () => {
     mock.restore();
 });
 
-test("standingOrders", async () => {
+test.skip("standingOrders", async () => {
```
Comment (Prior99) [inline packages/fints/src/__tests__/test-pin-tan-client.ts]: Please remove this `skip`.

File: packages/fints/src/client.ts [hunk 18]
```diff
@@ -35,6 +37,18 @@ export abstract class Client {
         ]));
         await dialog.end();
         const hispa = response.findSegment(HISPA);
+
+        hispa.accounts.map((account) => {
+            const hiupdAccount = dialog.hiupd.filter((element) => {
+                return (element.account.iban === account.iban);
+            });
+            if (hiupdAccount.length > 0) {
+                account.accountOwnerName = hiupdAccount[0].account.accountOwnerName1;
```
Comment (Prior99) [inline packages/fints/src/client.ts]: Instead of assigning to `account` here, mutating the input data, return a new object:

```
{
    ...acccount,
    accountOwnerName: hiupdAccount[0].account.accountOwnerName,
    accountName: hiupdAccount[0].account.accountName,
    limitValue: Parse.num(hiupdAccount[0].account.limitValue),
}
```

and the return the result of the `.map`.

File: packages/fints/src/client.ts [hunk 19]
```diff
@@ -92,39 +106,60 @@ export abstract class Client {
      *
      * @return A list of all statements in the specified range.
      */
-    public async statements(account: SEPAAccount, startDate: Date, endDate: Date): Promise<Statement[]> {
+    public async statements(account: SEPAAccount, startDate?: Date, endDate?: Date): Promise<Statement[]> {
         const dialog = this.createDialog();
         await dialog.sync();
         await dialog.init();
+        const segments: Segment<any>[] = [];
+        segments.push(new HKKAZ({
```
Comment (Prior99) [inline packages/fints/src/client.ts]: I have no insights into the domain of this, can you link to me some documentation?

File: packages/fints/src/client.ts [hunk 20]
```diff
@@ -92,39 +106,60 @@ export abstract class Client {
...
         await dialog.end();
-        const segments: HIKAZ[] = responses.reduce((result, response: Response) => {
+        const responseSegments: HIKAZ[] = responses.reduce((result, response: Response) => {
             result.push(...response.findSegments(HIKAZ));
             return result;
         }, []);
-        const bookedString = segments.map(segment => segment.bookedTransactions || "").join("");
-        const unprocessedStatements = await read(Buffer.from(bookedString, "utf8"));
+        const bookedString = responseSegments.map(segment => segment.bookedTransactions || "").join("");
+        const unprocessedStatements = await read(Buffer.from(bookedString, "ascii"));
```
Comment (Prior99) [inline packages/fints/src/client.ts]: Why did you change from "utf8" to "ascii"?

File: packages/fints/src/dialog.ts [hunk 21]
```diff
@@ -140,14 +159,22 @@ export class Dialog extends DialogConfig {
...
 
         const response = await this.connection.send(request);
         if (!response.success) {
-            const errors = response.errors.map(error => `"${error}"`).join(", ");
-            throw new Error(`Error(s) in dialog: ${errors}.`);
+            throw new ResponseError(response);
+        }
+        if (response.returnValues().has('0030')) {
+            const hitan = response.findSegment(HITAN);
+            throw new TanRequiredError(response.returnValues().get('0030').message,
```
Comment (Prior99) [inline packages/fints/src/dialog.ts]: Please fix  the formatting:

```
throw new TanRequiredError(
    response.returnValues().get('0030').message, 
    hitan.transactionReference,
    hitan.challengeText,
    hitan.challengeMedia,
    this,
);
```

File: packages/fints/src/http-connection.ts [hunk 22]
```diff
@@ -30,16 +30,19 @@ export class HttpConnection extends ConnectionConfig implements Connection {
 
     public async send(request: Request): Promise<Response> {
         const { url } = this;
+        console.log(`Sending Request: ${request}`);
```
Comment (Prior99) [inline packages/fints/src/http-connection.ts]: Remove this line.

File: packages/fints/src/http-connection.ts [hunk 23]
```diff
@@ -30,16 +30,19 @@ export class HttpConnection extends ConnectionConfig implements Connection {
...
         if (this.debug) { verbose(`Parsed Request:\n${request.debugString}`); }
         const httpRequest = await fetch(url, {
             method: "POST",
             body: encodeBase64(String(request)),
         });
         if (!httpRequest.ok) { throw new Error(`Received bad status code ${httpRequest.status} from FinTS endpoint.`); }
+
         const responseString = decodeBase64(await httpRequest.text());
-        const response = new Response(responseString);
+        console.log(`Received Response: ${responseString}`);
```
Comment (Prior99) [inline packages/fints/src/http-connection.ts]: Remove this line.

File: packages/fints/src/segments/hicdb.ts [hunk 24]
```diff
@@ -51,7 +51,7 @@ export class HICDB extends SegmentClass(HICDBProps) {
             interval: Parse.num(interval),
             orderDay: Parse.num(orderDay),
             lastOrderDate: lastOrder ? Parse.date(lastOrder) : null,
-            creationDate: Parse.date(jsonMessage.GrpHdr.CreDtTm),
+            creationDate: new Date(jsonMessage.GrpHdr.CreDtTm),
```
Comment (Prior99) [inline packages/fints/src/segments/hicdb.ts]: Why did you change this?

File: packages/fints/src/segments/hiupd.ts [hunk 25]
```diff
@@ -0,0 +1,33 @@
...
+			[accountNumber, subAccount, a, blz],
+			[iban],
+			[clientId],
+			[accountType],
+			[accountCurrency],
+			[accountOwnerName1],
+			[accountOwnerName2],
+			[accountName],
+			[limitType, limitValue, limitCurrency],
+			, ...parameters] = input;
```
Comment (Prior99) [inline packages/fints/src/segments/hiupd.ts]: FIx formatting (move closing square bracket into next line) please.

File: packages/fints/src/segments/hktan.ts [hunk 26]
```diff
@@ -1,42 +1,64 @@
...
+      } else {
+        if (version === 6) {
+          return [process, "HKIDN"];
+        } else {
+          return [process];
         }
+      }
+    } else if (process === "2") {
+      if (version === 6) {
+        return [process, "", "", "", aref, "N"];
```
Comment (Prior99) [inline packages/fints/src/segments/hktan.ts]: Please fix formatting in this file.

File: README.md [hunk 27]
```diff
@@ -1,10 +1,11 @@
-# fints
+# fints-lerna
```
Comment (Prior99) [inline README.md]: As this is the toplevel readme, I think this `-lerna` might be confusing.

File: package.json [hunk 28]
```diff
@@ -7,6 +7,9 @@
   ],
   "private": true,
   "devDependencies": {
-    "lerna": "^3.2.1"
-  }
+    "lerna": "^3.19.0",
+    "tslint": "^5.20.0",
+    "typescript": "^3.7.3"
+  },
+  "name": "fints-lerna"
```
Comment (Prior99) [inline package.json]: I am unsure whether this package really needs a name. It's not intended to be published.

File: package.json [hunk 29]
```diff
@@ -7,6 +7,9 @@
   ],
   "private": true,
   "devDependencies": {
-    "lerna": "^3.2.1"
-  }
+    "lerna": "^3.19.0",
+    "tslint": "^5.20.0",
```
Comment (Prior99) [inline package.json]: Are tslint and typescript really needed in the toplevel lerna package?

File: packages/fints/src/segments/hicdb.ts (hunk 24 above)
Comment (mathiasarens) [reply packages/fints/src/segments/hicdb.ts]: Screenshot 2020-02-16 at 09 27 46

File: packages/fints/src/segments/hiupd.ts (hunk 25 above)
Comment (mathiasarens) [reply packages/fints/src/segments/hiupd.ts]: done

File: packages/fints/src/__tests__/test-pin-tan-client-e2e.ts (hunk 6 above)
Comment (mathiasarens) [reply packages/fints/src/__tests__/test-pin-tan-client-e2e.ts]: done

**Comment by mathiasarens** [comment]
I updated the code according to your suggestions. But not all of the issues are resolved. The test-pin-tan-client-e2e.ts must be moved to the fints-cli package probably. The test was a very helpful end-to-end test during development. Please have a look at the changes.

File: packages/fints/src/__tests__/test-pin-tan-client-e2e.ts (hunk 7 above)
Comment (mathiasarens) [reply packages/fints/src/__tests__/test-pin-tan-client-e2e.ts]: Maybe I must move this 'test' to the fints-cli package. I used it during development.

File: packages/fints/src/segments/hktan.ts (hunk 26 above)
Comment (mathiasarens) [reply packages/fints/src/segments/hktan.ts]: This is the standard formatting from Visual Studio

File: packages/fints/src/__tests__/test-pin-tan-client-e2e.ts (hunk 7 above)
Comment (Prior99) [reply packages/fints/src/__tests__/test-pin-tan-client-e2e.ts]: I'd suggest to create a new type of tests (perhaps in a different sub-package?) called E2E tests. We could move this kind of tests in there and only execute them on demand. What do you think?

File: packages/fints/package-lock.json (hunk 4 above)
Comment (mathiasarens) [reply packages/fints/package-lock.json]: Removed

File: packages/fints/src/__tests__/test-pin-tan-client-e2e.ts (hunk 7 above)
Comment (mathiasarens) [reply packages/fints/src/__tests__/test-pin-tan-client-e2e.ts]: That's a good idea, too. But I noticed that the fints-cli is broken with the new fints module. I want to repair this too. And this e2e-test is indeed a fints-cli executions. Having said that, I think this test should be removed.

File: README.md [hunk 30]
```diff
@@ -1,7 +1,6 @@
 # fints
 
-[![pipeline status](https://gitlab.com/prior99/fints/badges/master/pipeline.svg)](https://github.com/Prior99/fints)
-[![coverage report](https://gitlab.com/prior99/fints/badges/master/coverage.svg)](https://github.com/Prior99/fints)
+[![npm](https://img.shields.io/npm/v/node-fints.svg)](https://www.npmjs.com/package/node-fints)
```
Comment (Prior99) [inline README.md]: Please revert the badges in the readme to the way they were before.

File: package.json (hunk 28 above)
Comment (Prior99) [reply package.json]: I believe the package doesn't need a name, please remove it.

File: packages/fints/README.md [hunk 31]
```diff
@@ -1,13 +1,16 @@
 # fints
 
-[![npm](https://img.shields.io/npm/v/fints.svg)](https://www.npmjs.com/package/fints)
-[![pipeline status](https://gitlab.com/prior99/fints/badges/master/pipeline.svg)](https://github.com/Prior99/fints)
-[![coverage report](https://gitlab.com/prior99/fints/badges/master/coverage.svg)](https://github.com/Prior99/fints)
+[![npm](https://img.shields.io/npm/v/node-fints.svg)](https://www.npmjs.com/package/node-fints)
```
Comment (Prior99) [inline packages/fints/README.md]: Please revert the badges to the way they were before.

File: packages/fints/README.md [hunk 32]
```diff
@@ -1,13 +1,16 @@
...
 
 A client library for communicating with [FinTS servers](https://www.hbci-zka.de/).
 
+Fork from [fints](https://github.com/Prior99/fints)
+
+Fork from [nodejs-fints](https://github.com/as19git67/nodejs-fints)
```
Comment (Prior99) [inline packages/fints/README.md]: Please remove these lines. this is not a fork.

File: packages/fints/src/__tests__/test-pin-tan-client-e2e.ts (hunk 7 above)
Comment (Prior99) [reply packages/fints/src/__tests__/test-pin-tan-client-e2e.ts]: OK

File: packages/fints/src/client.ts [hunk 33]
```diff
@@ -35,6 +37,21 @@ export abstract class Client {
...
         const hispa = response.findSegment(HISPA);
+
+        hispa.accounts.map((account) => {
+            const hiupdAccount = dialog.hiupd.filter((element) => {
+                return (element.account.iban === account.iban);
+            });
+            if (hiupdAccount.length > 0) {
+                return {
+                    ...account,
+                    accountOwnerName: hiupdAccount[0].account.accountOwnerName1;
```
Comment (valoricDe) [inline packages/fints/src/client.ts:48]: did this compile on your side?

File: packages/fints/src/client.ts [hunk 34]
```diff
@@ -92,39 +109,60 @@ export abstract class Client {
...
+        const segments: Segment<any>[] = [];
+        segments.push(new HKKAZ({
+            segNo: 3,
+            version: dialog.hikazsVersion,
+            account,
+            startDate,
+            endDate
+        }));
+        if (dialog.hktanVersion >= 6) {
+            segments.push(new HKTAN({ segNo: 4, version: 6, process: "4", segmentReference: "HKKAZ", medium: dialog.tanMethods[0].name }));
```
Comment (valoricDe) [inline packages/fints/src/client.ts:125]: did you have TSLint installed?
